packages/opentelemetry-tracing/src/platform/browser/export/BatchSpanProcessor.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { BatchSpanProcessorBase } from '../../../export/BatchSpanProcessorBase';
+import { SpanExporter } from '../../../export/SpanExporter';
+import { BatchSpanProcessorBrowserConfig } from '../../../types';
+
+export class BatchSpanProcessor extends BatchSpanProcessorBase<BatchSpanProcessorBrowserConfig> {
+  private _visibilityChangeListener?: () => void
+
+  constructor(_exporter: SpanExporter, config?: BatchSpanProcessorBrowserConfig) {
+    super(_exporter, config)
+    this.onInit(config)
+  }
+
+  private onInit(config?: BatchSpanProcessorBrowserConfig): void {
+    if (config?.disableAutoFlushOnDocumentHide !== true && document != null) {
+      this._visibilityChangeListener = () => {
```

Comment:
This should also trigger flush on a `pagehide` event - Safari doesn't fully support `visibilitychange`: https://caniuse.com/?search=visibilitychange

Reply:
Good catch. I added support for the `pagehide` event.
I didn't add any check to prevent double call of `forceFlush` when both `pagehide` and `visibilitychange` events are dispatched to make the code simpler. Let me know if you want such mechanism.

Reply:
I don't think there's any need for that level of complexity - `flush` implementations should be smart enough.

Reply:
smart enough to be called twice concurrently you mean?

Reply:
There is also ```beforeunload``` and ```unload``` (depending on the browser support matrix that we want to support -- specifically older IE instances (which are probably already excluded by other usages)) that you may want to also consider adding.

Additionally, there is actually a set of corner cases that we should support around pages hooking theses same events after the SDK has initialized but still creating / sending telemetry. The simplest solution for this is to set a flag during these flush events (or more specifically the unload events) and then during the onEnd() as well as batching you also trigger the same auto flush.

Reply:
I described my concerns about using `unload` event here https://github.com/open-telemetry/opentelemetry-js/issues/2205 .

Also I was testing JS auto-instrumentation on different browsers and on IE11 I got a `Script error` so I guess we don't support it.

This PR is about auto flushing created spans when page become hidden, so I don't think there is something wrong with sending telemetry when page is not active, or I didn't get your point right?

Reply:
-1 for `unload`: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon has a full writeup of the recommended approach.  What's in the PR now looks good.  

If the current `BatchSpanExporter.forceFlush` isn't smart enough to deal with two calls in succession, I think it should be improved as a separate PR rather than complicating state/logic here.  A surface skim looks like it might have some issues...

Reply:
My comment is not a blocking comment for this PR -- as @johnbley says What's in the PR now looks good as it's making this better